csharp/src/Google.Protobuf/Reflection/FieldDescriptor.cs

```diff
@@ -90,6 +90,12 @@ public sealed class FieldDescriptor : DescriptorBase, IComparable<FieldDescripto
         /// </summary>
         public override string Name { get { return proto.Name; } }
 
+
+        /// <summary>
+        /// The json_name option of the descriptor's target.
+        /// </summary>
+        public string JsonName { get { return proto.JsonName; } }
```

Comment:
Any reason not to make this always return "the name to use in JSON"? I'd implement it as:

```csharp
public string JsonName => proto.JsonName == "" ? Name : proto.JsonName;
```

(I need to get round to C# 6-ifying the rest of the properties in here at some point. Feel free to use the more longwinded syntax for now.)

At that point, the rest of this CL can just change to use that property unconditionally, which will simplify things.

Note that we can just check for "" as a string property in a proto will never have a value of null.